### ISSUE TITLE
As a developer, I want to identify and implement a solution for event-based evaluations

### ISSUE THREAD

**Issue by epag** [NOAA-OWP/wres#130]
---

Author Name: **Hank** (Hank)
Original Redmine Issue: 119823, https://vlab.noaa.gov/redmine/issues/119823
Original Date: 2023-08-22

---

See #79741 for the requirement being addressed.

This ticket can be resolved once we have designed a solution and implemented it.

Leaving as normal and in the backlog pending prioritization of the NWM team requirements for use case #115608.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-08-22T14:10:58Z

---

Going high with my estimated time.  Will wait to see how James adjusts it,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **James** (James)
Original Date: 2023-08-22T14:30:24Z

---

This is another ticket where it depends hugely on the starting assumptions.

If I assume that the method for event detection (defined as some function that consumes a time-series and returns a set of non-overlapping datetime intervals that correspond to "events") is one of:

* The method used by the NWM team (Yuqiong et al.); or
* The method used by ISED in their hydrotools (Jason).

Then my estimate for a first cut is probably closer to "64", but let's say "128".

Conversely, if we're starting from scratch, reviewing and perhaps even inventing, then my estimate is much higher, probably 2^10=1024.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **James** (James)
Original Date: 2023-08-22T14:32:30Z

---

Oh, and I would like to add "no forecasts" to the starting assumption above. I'm not sure how event detection is done for forecasts, which will quite often not capture an event in its entirety. I haven't looked at either of the above two methods in detail, but I assume neither admit forecasts.

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-08-22T15:06:54Z

---

I'll assume the simple solution, for now, and lower the estimate. I would think we would want to build off of either the ISED or NWM algorithm instead of starting from scratch.  

Jay emphasized that the schedule can always be adjust after the reauthorization if its determined that an estimate was off.  

Hank

**Comment by james-d-brown** [comment]
Marking this as nominally started, but will probably begin with a conversation in the original user support ticket over on VLab.

**Comment by james-d-brown** [comment]
Kicked off a conversation in the relevant ticket. Will wait to see if anyone responds.

**Comment by james-d-brown** [comment]
No responses so far. I do wonder whether anyone really wants this feature. For that reason, I lean towards something quite minimal. I will follow-up with Jason R. and ask about the method they use in hydrotools for event detection.

**Comment by james-d-brown** [comment]
Sent a follow-up to Jason R.

**Comment by james-d-brown** [comment]
Going to meet with Jason R to discuss further.

**Comment by james-d-brown** [comment]
Regarding the event detection method, it is likely to be a bit hand-wavey in the sense that it must be an automated scheme because the WRES is not an interactive program, although some parameters are likely to be needed (with defaults). Regardless, the choice of parameters introduces a hand-wavey element and, depending on the chosen technique, there may be some fundamental limitations in some circumstances. 

Some discussion was had about the extent to which discrete frequencies could be identified and modeled, like a low-frequency trend or periodic signal (e.g., seasonal), a high-frequency periodic signal (e.g., diurnal), a high-frequency non-periodic signal and a high-frequency non-structural or noise signal. The following paper adopts such an approach:

Regina, J.A., F.L. Ogden, 2021.  Automated Correction of Systematic Errors in High-Frequency Depth Measurements above V-Notch Weirs using Time Series Decomposition. Hydrol. Proc.

With code:

https://github.com/NOAA-OWP/hydrotools/blob/main/python/events/src/hydrotools/events/event_detection/decomposition.py

Naturally, there will be some overlap between these categories and some modeling challenges in some circumstances. For example, in a hydrologic context, the extent to which rainfall-driven runoff (high-frequency non-periodic) could be discriminated from periodic diurnal variations (high-frequency periodic) may be complicated when rainfall itself is partly diurnal (e.g., evening convection). Nevertheless, this and most other techniques can only provide an approximate discrimination of event start and end dates and they should not be regarded as precise. In some cases "what is an event?" doesn't have a clear answer. For example, in snow basins, the melt period may be hard to distinguish as a discrete event.

Furthermore, it is likely that some post-processing parameters may be needed to filter out events that are too short or do not represent a sufficient rise (e.g., above a threshold). Again, hand-wavey parameters, but they are likely to be unavoidable.

To the extent that suitable methods exist that address different types of event discrimination, it should be possible to support a method designation/declaration with up to N methods of event discrimination supported, and a default imposed in the absence of a selection.

**Comment by james-d-brown** [comment]
Some further discussion was had about how to model/separate baseflows from the runoff signal and, for example, it was suggested that a digital recursive filter may be adequate in many circumstances, along the lines of:

Eckhardt, K. (2005). How to construct recursive digital filters for baseflow separation. Hydrological Processes, 19(2), 507–515. https://doi.org/10.1002/hyp.5675

In general, such filters assume a linear reservoir model and are, therefore, likely to work well in circumstances where the underlying flow generating mechanisms are approximately linear.

In short, regardless of the approach taken to discriminating "events" - these being high-frequency non-periodic variations roughly corresponding to rainfall-driven runoff in hydrologic terms - the actual event periods will be substantially approximate and the techniques and parameters applicable in particular circumstances and requiring some experimentation and guidance for users.

**Comment by jarq6c** [comment]
@james-d-brown These comments seem like a fair representation of the intricacies and pitfalls of event-based analysis. I'd add some additional notes for consideration.

> we can only accommodate long time-series of predictions like simulations because event detection with forecasts is unlikely to work

It's true that event detection and baseflow separation techniques tend not to work well (or at all) on short time series. This is because you need a signal long enough to characterize the trend you're trying to remove. From this perspective, "events" are really a characteristic of the time series as a whole ("bumps").

With regard to forecasts, I've wondered if event detection might work on lead time pools / time-lagged ensembles. You could generate a long single trace by taking the mean (or other statistic) across the same lead time pools of several forecasts. This could exacerbate the "time series component" identifiability problem if stringing together lead time pools results in more "bumps" (as it would if the forecast system uses anything like The Nudge™). Nonetheless, it is a way to get a forecast trace long enough to model a trend.

> Regarding the event detection method, it is likely to be a bit hand-wavey in the sense that it must be an automated scheme because the WRES is not an interactive program, and some parameters are likely to be needed (with defaults).

We've still yet to conduct a sensitivity analysis of event detection. I've heard anecdotally from users that some parameters don't seem to really change the identified events. It might be interesting to try and quantify how hand-wavey we can get.

> The following paper adopts such an approach, modeling the overall signal as a sum of parts:

I'd note here that we chose the additive model because this model of time series decomposition is implicitly assumed by baseflow separation techniques. That is many conceptualizations of hydrology implicitly assume discrete components of streamflow originate from discrete physical processes in a catchment each of which additively contributes a discrete volume to the total streamflow according to some conservation equation.

From a purely signal and systems perspective, we could have used a multiplicative or exponential model of composition, but I don't think I've seen anything other than an additive model used in hydrology. I suppose I'm just re-emphasizing your point that the major source of uncertainty here is the appropriate number of time series "parts" and how these parts vary in time.

> Eckhardt, K. (2005). How to construct recursive digital filters for baseflow separation. Hydrological Processes, 19(2), 507–515. https://doi.org/10.1002/hyp.5675

You may find these papers interesting. The Eckhardt filter has two parameters: recession constant and maximum baseflow index. These papers discuss automated techniques for setting these parameters. 

1. This paper presents a method for using recession analysis to automatically determine the recession constant.

_Eckhardt, K. (2008). A comparison of baseflow indices, which were calculated with seven different baseflow separation methods. Journal of Hydrology, 352(1-2), 168-173._

2. This paper uses a reverse filtering technique to automatically determine the maximum baseflow index.

_Collischonn, W., & Fan, F. M. (2013). Defining parameters for Eckhardt's digital baseflow filter. Hydrological Processes, 27(18), 2614-2622._

3. This paper discusses sensitivity and parametric uncertainty of the filter.

_Eckhardt, K. (2012). Analytical sensitivity analysis of a two parameter recursive digital baseflow separation filter. Hydrology and Earth System Sciences, 16(2), 451-455._

**Comment by james-d-brown** [comment]
Much appreciated, thanks, Jason!

On forecasts, I agree that we should consider forecasts, probably as an extension of the initial work. For long-range forecasts, it should be mechanically possible and, even for short- or medium-range forecasts, it may be possible to look across multiple issued datetimes to build a picture of events as seen by (multiple) forecasts. Also, as you say, it should be possible to look across lead time pools, building a long time-series for a fixed lead time, repeated for all lead times. Anyway, that sounds like interesting follow-up work.

**Comment by jarq6c** [comment]
Another consideration I forgot about is compound events. Depending on the hydrologist, the attached image might contain 1 multi-day event or 8ish individual event peaks. I imagine most definitions would land on 1, 3, or 5 events, but that's not much clearer.

![Screenshot from 2024-11-20 10-29-32](https://github.com/user-attachments/assets/d9e45885-b83d-492f-bffd-975d6561a200)

**Comment by james-d-brown** [comment]
Yes, for sure, that is part of the subjectivity of it all, when does an event start and end? There cannot really be a universally "right" answer to that because it partly depends on the question being posed. I think the best we can do is to allow a user to tweak some hand-wavey post-processing parameter, like a minimum gap between events before aggregation (into one event). That said, the way I see this ticket to begin with is to provide our users with a starting point. If it generates some interest and usage, more will be done. If it doesn't, well, more will not be done :)

**Comment by james-d-brown** [comment]
A related point in this context is the architecture. We originally discussed an event detection service. I still think that makes more sense, ultimately, especially if a user needs to play with parameters before they get the events they want, visually/subjectively speaking. That process should be independent of the actual use of the events in an evaluation, and it doesn't make much sense to run an evaluation tool as an event detection tool (without an evaluation bolted on). Separation of concerns and all that.

**Comment by james-d-brown** [comment]
Back to this.

**Comment by james-d-brown** [comment]
One nice thing about the `covariate` context is that a user can potentially declare entirely different variables for event detection than the variables used for evaluation. Whether this makes sense is entirely context dependent, but it is flexible and desirable.

**Comment by james-d-brown** [comment]
The other obvious benefit of re-using the existing declaration for datasets is that the dataset declaration drives an existing data reading and ingest pipeline, both of which are pre-requisites for event detection. In other words, we don't simply re-use declaration, but a large chunk of our existing infrastructure for getting time-series data into the software.

**Comment by james-d-brown** [comment]
However, once we move beyond ingest, things become more complicated and will require a significant amount of new development.

Recall that the phase of our pipeline after ingest is entirely "pool-shaped". That is to say, data retrieval, upscaling, pairing etc. all takes place in the context of a pre-defined pool, repeated in parallel for each pool in the evaluation. A pool is described by a time window or temporal "event" boundaries, among other things. In other words, a pool is effectively event-shaped and we will simply re-use this concept for event-shaped evaluations based on event detection, rather than explicit declaration.

This all sounds good/fine, but the obvious issue here is that we cannot start that part of our existing pipeline that is event-shaped without the events to drive them. Since we are introducing pools ("events") that are entirely data-driven, the chicken and the egg are in conflict here, i.e., we need to detect the pools before we can use them to drive our post-ingest pipeline.

In short, we will need a new software infrastructure after the ingest phase and before the evaluation phase that is concerned with generating the pool boundaries or time windows for event-based evaluations, which will go on to drive the existing post-ingest pipeline.

We can still re-use lots of things, like time-series retrievers and upscalers and so on, because our software is properly abstracted. However, we cannot incorporate event detection into the pool-shaped pipeline that already does all of these activities, because event detection is concerned with identifying the pools in the first place. Also, for the avoidance of doubt, this isn't a problem with our software design - it makes perfect sense for the post-ingest work to be pool-shaped. But we do need new abstractions to retrieve and upscale etc. the time-series data for event detection and then to perform the event detection itself.

**Comment by james-d-brown** [comment]
All of that to say, even the first cut at this is a big ticket in terms of integration alone, putting aside any time to develop a sensible event detection algorithm. I have copied across the estimate of 128 bananas from the Redmine ticket, which was dropped from 256 bananas, but the reality is probably somewhere in between these two estimates and the provisional deadline of mid Jan '25 may need to slip given the integration work. However, we can clarify that as things progress.

**Comment by james-d-brown** [comment]
Back to the declaration, I expect the simplest possible declaration could look something like this:

```
observed: observations.csv
predicted: predictions.csv
event_detection: observed
```

This would be a shorthand for:

```
observed: observations.csv
predicted: predictions.csv
event_detection: 
  dataset: observed
```

In other words, conduct an evaluation using all default settings and event detection with the `observed` data source (and using all default settings for the event detection).

**Comment by james-d-brown** [comment]
To use a difference dataset for event detection, it could look something like this:

```
observed: observations.csv
predicted: predictions.csv
covariates:
  - sources: covariates.csv
    purpose: detect
event_detection: covariates
```

Again, the `purpose` qualifier would be to distinguish the use of this covariate for event detection and not for default filtering of the pairs. It is slightly more cumbersome than ideal, as the `covariates` could otherwise be declared in the same was as the `observed` or `predicted`, but this maintains backwards compatibility and, ultimately, it is more likely that covariates will be used for filtering than event detection.

**Comment by james-d-brown** [comment]
I think the more complex declarations flow naturally from that and there probably isn't a need to elaborate further at this stage. In other words, we can assume that `event_detection` will have a bunch of optional parameters, but only the `dataset` as a required parameter, declared in long or short form, but always as an orientation or reference to one of the formally declared datasets.

**Comment by james-d-brown** [comment]
It is perhaps worth clarifying that both the `dataset` and the `purpose` could be a list. In other words:

```
observed: observations.csv
predicted: predictions.csv
covariates:
  - sources: covariate.csv
    purpose: 
      - detect
      - filter
event_detection: 
  datasets:
     - observed
     - covariates
```

In other words, use the covariate for both filtering and event detection and perform event detection using both the `observed` and `covariates` datasets. Perhaps improbable, but perfectly possible.

**Comment by james-d-brown** [comment]
I will most likely proceed with sketching out the integration work in code and leave the event detection algorithm until later. The latter can be mocked with a placeholder that returns the unbounded time window, equivalent to the universal set. The integration work is the hard part, at least in terms of software engineering.

**Comment by james-d-brown** [comment]
I don't think it's really an option to declare the datasets separately within the context of event detection as that would both lead to duplication for the 99% case where the event detection is based on the `observed` dataset (most likely) and it would also drive a ton of duplication under the hood because we'd need to adjust pathways for getting the data into the software. In short, that is a no go.

I appreciate your concern about the special qualifier when a `covariate` is used, as I had the same concern, but we need to remember two things here. First, declaring the dataset for event detection as a `covariate` is not likely to be the majority use case. Second, when a covariate has more than one purpose, it stands to reason that the purpose is declared explicitly, and I don't think we can change the default situation where a covariate is used as a filter.

Or, rather, we should at least consider the consequences of the latter. For example, let's say that we interpret a `covariate` as a filter when either a `minimum` or a `maximum` is declared, otherwise it's intended for event detection. That's fine and understandable, but it leads to some significant downsides, namely that a covariate cannot serve more than one purpose at once, that a covariate cannot ever serve more than two purposes (because then we'd need an explicit choice anyway) and that we introduce a breaking change too.

So, overall, I don't think this is much of a decision. The only decision is about the phrasing of the declaration, I think.

**Comment by HankHerr-NOAA** [comment]
That all makes sense. Thanks,

Hank

**Comment by james-d-brown** [comment]
Have the declaration sketched out. Need to add some validation, notably against `valid_dates` in combination with `event_detection`, but probably some warnings too, such as when other pool sequences are declared. Also need some unit tests.

**Comment by james-d-brown** [comment]
Running the system tests and will then push progress.

**Comment by james-d-brown** [comment]
Progress in b64848130e71e2f711a0c5ccac9b0d165f9485d6. More next week.

**Comment by james-d-brown** [comment]
Have added some declaration validation.

**Comment by james-d-brown** [comment]
Have the placeholder wiring done, together with the declaration aspects, at least provisionally. Will push progress shortly.

The next major challenge (first major challenge with wiring) concerns the relationship between geographic features and events and how that may constrain the user stories that are supported initially. Some thoughts on that forthcoming.

**Comment by james-d-brown** [comment]
First, some background or context. The software design or evaluation pipeline is currently organized around the calculation of statistics for "pools". Recall that a pool is an atomic unit of work or an atomic unit for which statistics are calculated.

A pool contains the time-series data for:

1. One or more geographic features; and
2. A time window, akin to an "event".

Other filters may be applied on top of pools (i.e., are applied to all pools equally), such as season filters and value filters.

The concept of a time-window will be used to drive event-based evaluations where the time windows are generated through data-driven event detection, just like evaluations where a user declares the time window explicitly or declares a sequence of time windows. Afterall, we are not going to reengineer the software to accommodate event detection when it uses fundamentally the same abstraction of time as existing evaluations, aka a time window. For event detection, the time dimension for detection is always valid time, although constraints on reference times and lead times may be declared and layered on top of all detected events. For example, "as a user, I want to evaluate timing errors for hydrologic events when the forecast lead times are between t and t+x".

In summary, a pool encapsulates a 1:1 mapping between a time window and a collection of geographic features (one or more). A pool typically contains one geographic feature, but it may contain several where a user aims to pool pairs across geographic features. As such, a challenge arises in terms of supporting user stories that involve pooling time-series data or summarizing statistics across geographic features in the context of event detection. Each event is a separate time window and there is a separate collection of time windows for each geographic feature.

**Comment by james-d-brown** [comment]
Consider the following user stories that involve multiple geographic features:

1. "As a user, I want to calculate the timing error of the model predictions for each hydrologic event and summarize these timing errors across geographic features for fast responding basins";
2. "As a user, I want to calculate an event-based Brier skill score across all geographic features";
3. "As a user, I want to calculate an event-based fractional bias of the model predictions across all geographic features associated with the hydrologic event of Hurricane Foo".

These user stories present different challenges in terms of implementation. 

Consider user story (1). This would involve performing event detection for each hydrologic basin (geographic feature) that has been nominated as fast-responding, calculating the timing error associated with each event (per feature) and then summarizing these timing errors across events and features, perhaps with a histogram or other summary statistics. This should be doable within the current design because the 1:1 mapping between a time window and a geographic feature is retained. The time-series data is gathered for each event (time window) and feature separately, statistics are calculated from these pools and then, finally, summary statistics are calculated across all pools. It would require that we support calculating summary statistics across all pools where we currently only support summary statistics across geographic feature groups or all geographic features for each time window separately, not across all time windows. Anyway, this user story is doable.

Consider user story (2). This is much harder to support. Calculating an "event-based BSS" implies that we have a pool with multiple geographic features, each of which has a separate collection of time-windows (events). This is contrary to the idea that a pool abstracts a single time window (event). Is this even a sensible user story? Probably not. Pooling time-series data across geographic features and events will rarely make sense in a hydrological context, at least, but it remains a *potential* user story.

Consider user story (3). This is also challenging, potentially. In practice, the way this has been achieved to date is for a user to nominate an overall event period, i.e., a common time window across all geographic features. For example, Hurricane Helene occurred in late September 2024, meteorologically speaking. Of course, this doesn't mean that, hydrologically speaking, its effects had dissipated by the end of September. On the contrary, the hydrologic timing will vary across basins. Still, a user can roughly capture an event period of interest in this way. However, if they wanted to perform event detection, the same challenge emerges with user story (2), namely that a pool is abstracted as containing a single time window (event) per feature group (of one or more features).

Most likely, the way a user would practically achieve stories like (2) and (3) using WRES would be to calculate summary statistics, the same as user story (1). In other words, they would calculate statistics per-feature-per-event and then summarize them across all events and features (or all events per-feature) using an appropriate summary statistic like a mean, i.e., a feature-and-event-averaged bias fraction or a feature-and-event-averaged-BSS. Statistically speaking, this probably makes more sense too because pooling pairs across features carries the risk of outliers having undue influence where basin hydrology varies markedly.

To achieve complete flexibility, a pool that abstracts several geographic features would need to support feature-specific time windows, but this is massively complicating and is unlikely to be supported any time soon.

**Comment by james-d-brown** [comment]
In summary, we will support evaluations that contain multiple geographic features with a single geographic feature in each pool, alongside event detection. For each feature, events will be detected separately and the associated time windows will drive time-series data gathering and calculation of statistics for each feature. We will also support summary statistics across all time windows (events), possibly in the first iteration of this ticket.

However, when it comes to evaluations that involve calculating events per feature and pooling the pairs across features, we will not support those evaluations because this is contrary to our fundamental description of a pool as containing a single time window. In short, feature groups with more than one feature per group will not be supported alongside event detection.

**Comment by james-d-brown** [comment]
Will assert these constraints upfront in terms of declaration validation.

**Comment by james-d-brown** [comment]
Also need to add some validation against forecast sources for event detection, both before ingest (when `type` has been declared) and after ingest (when `type` has been detected).

**Comment by james-d-brown** [comment]
Extra validation and unit tests added for forecast datasets, as well as datasets that don't exist.

**Comment by james-d-brown** [comment]
Have the basic wiring done, although I anticipate some issues and it needs integration testing. The latter is hard to do without realistic assertions, i.e., expected events for given series, which in turn requires an actual detection algorithm. Will run the system tests and then push progress. Step after that is to generate an events detection package, which contains a helper that performs the actual detection. Currently, there is a placeholder at the end of the wired chain that returns the universal time window.

**Comment by HankHerr-NOAA** [comment]
James,

I reviewed the comments in which you describe the issue of event detection with multiple features. It all seems clear to me and I agree with https://github.com/NOAA-OWP/wres/issues/130#issuecomment-2531541285 where you explain what we will support ("evaluations that contain multiple geographic features with a single geographic feature in each pool, alongside event detection", and summary statistics) and we won't support ("evaluations that involve calculating events per feature and pooling the pairs across feature").  All sounds reasonable to me,

Hank

**Comment by james-d-brown** [comment]
Good, thanks.

**Comment by james-d-brown** [comment]
Added and pushed a new module for event detection, which contains an API and a placeholder implementation. The next heavy lift is to implement an event detection algorithm, specifically the one described by Regina and Ogden (2021), https://doi.org/10.1002/hyp.14405.

More on that next week.

**Comment by james-d-brown** [comment]
Working on porting this across:

https://github.com/NOAA-OWP/hydrotools/blob/main/python/events/src/hydrotools/events/event_detection/decomposition.py

**Comment by james-d-brown** [comment]
Adjusting the API to consume a window size in duration units for smoothing, rather than an integer number of timesteps, as I think a duration is more intuitive, especially given that the other parameters are durations.

In terms of parameterization, there is one small difference with the Python implementation that follows the guidance in the Python docs, namely that the minimum event duration should typically be the half life for the exponential decay/smoothing when undefined. The python code defaults to zero, but the Java implementation defaults to the documented guidance.

**Comment by james-d-brown** [comment]
Going to run the system tests and then push progress.

**Comment by james-d-brown** [comment]
Progress pushed. Next step is to test with some realistic datasets. This part will take a while and will probably uncover some issues... more on that tomorrow.

**Comment by james-d-brown** [comment]
Identified a couple of issues with the first test of real data containing NaNs. First, when setting the default values of a couple of the parameters, the (average) timestep of the data is inspected and it's better to use the mode rather than the mean in the presence of missing data (missing timesteps), so adjusted that. Also, found an error in the handling of NaN values during exponential smoothing and fixed that.

**Comment by james-d-brown** [comment]
Pushed those changes.

**Comment by james-d-brown** [comment]
Weird error generating/submitting the dependency graph. Rerunning that build step...

**Comment by james-d-brown** [comment]
OK, succeeded this time. Looks like an http server error of some kind.

**Comment by james-d-brown** [comment]
Might be useful to have a visual representation of the detected events, i.e., a metric and associated graphic. Will consider that later. Currently, the events are logged, as time windows are logged.

**Comment by james-d-brown** [comment]
> Might be useful to have a visual representation of the detected events, i.e., a metric and associated graphic. Will consider that later. Currently, the events are logged, as time windows are logged.

Would probably make sense to add a metric to support this as it would allow for event exploration/refinement prior to expanding the evaluation with other metrics. However, this would be a *lot* of work as we don't have a comparable statistic class. It would need to store the paired time-series data, plus the event boundaries (two instants). For that reason, it is probably beyond the scope of this initial ticket. TBD.

**Comment by james-d-brown** [comment]
As noted earlier, when detecting events across multiple source types (e.g. observed and predicted), then the treatment of events could follow either the union, i.e., use events present in any source, or the intersection, i.e., use events only common to all sources. Also, the comparison for equality could involve various modes, but overlap is probably the most sensible choice. Currently, the default is to form the union and there is no declaration option or logic to perform an intersection, but we'll probably want to include that possibility in the initial implementation.

**Comment by james-d-brown** [comment]
Oops, system tests failed on assertions related to ordering of pools. Refactored the generation of time windows and forgot to return in a sorted set.

**Comment by james-d-brown** [comment]
Passing again.

**Comment by james-d-brown** [comment]
Adding support for the intersection of events across source types.

**Comment by james-d-brown** [comment]
> Adding support for the intersection of events across source types.

Done this. Need to add some integration tests of the `EventsGenerator` where this logic occurs - which is not a trivial effort.

**Comment by james-d-brown** [comment]
Working on integration testing of the `EventsGenerator`.

**Comment by james-d-brown** [comment]
Added some integration tests, which revealed a bug - fixed that.

Next and more arduous task is to experiment further with some real datasets.

There are a lot of parameters and event detection interacts with other features too, so there's quite a lot to test.

As to whether the detection itself is realistic, that is more about the algorithm and the choice of parameters, but initial evidence is that this will require a lot of handholding, 

As noted earlier, the better long-term strategy is a separation of concerns between event detection and evaluation so that events can be detected and refined before the evaluation pipeline begins. 

In the interim, there will probably need to be an evaluation metric that performs (and provides the basis for visualization of) the event detection and thereby allows users to experiment with, and refine, the detected events, before starting an evaluation proper with event-shaped statistics. However, this is a somewhat ugly mitigation, conceptually, to a lack of separation of concerns and probably also beyond the scope of the initial implementation.

**Comment by james-d-brown** [comment]
Ah, before continuing with testing.

Regarding the combination strategy, which currently supports a union or intersection of events across time-series. When forming a union, the outcome seems pretty straightforward/unambiguous. However, when forming a temporal intersection, a question arises about whether to simply accumulate the events from all series (the default) or to allow for event aggregation. 

For example, say we get one event from one series (e.g., an observed series) that spans 1 January through 10 January and one event from another (e.g., simulated/predicted) series that spans 3 January through 15 January and the intersection is requested. In that case, with the default behavior, evaluation statistics will be produced for each of these two events separately, because they intersect and, indeed, there is no difference between a union and intersection in this case.

But perhaps a user might like to enforce an aggregation policy when an intersection occurs. For example, perhaps the user wants statistics for a single event from 1 January through 15 January (the maximum period spanned by the two intersecting events) or a single event from 2 January through 12Z on 12 January (the average of the start and end times of the two event periods) or a single event from 3 January through 10 January (the minimum period spanned by the two intersecting events or the period of temporal intersection).

Although this is arguably beyond the scope of the present ticket, it is likely to be quite useful/necessary when considering events across multiple data sources, so I lean towards supporting it.

**Comment by james-d-brown** [comment]
Have that additional functionality drafted.

The user documentation of this event detection capability is going to be somewhat of a pain... oh well, that comes later.

**Comment by james-d-brown** [comment]
Still no fix to the annoying indriya issue with an uncaught exception, logged to standard error:

https://github.com/unitsofmeasurement/uom-lib/issues/80

Oh well, not material.

**Comment by james-d-brown** [comment]
Having played with some real and synthetic datasets, it basically seems to work, although I'm not certain how practical it will be in reality. As noted earlier, there is a material sensitivity to the choice of parameter values and it is rather hard to identify good defaults. The choice of parameter values can make the difference of an order of magnitude or more in the number of events that are detected.

I suppose the upside of having many parameters is that it does provide some flexibility to nudge the detection in a particular direction, given the time to experiment and hand-hold it.

Here's an example to demonstrate.

Generate a sine wave with a length of 365 days, sampled every 0.1 day and with an amplitude of 5, arbitrarily, and a range of [0,10], and adding some arbitrary datetimes and other metadata to produce a wres csv input, like so:

```
track_length <- 365
upsample <- 10
rate1 <- 0.02
rate2_range <- list(0.00, 10)
x1 <- 1:(track_length*upsample)/upsample
amp <- (rate2_range[[2]] - rate2_range[[1]])/2
y1 <- amp*cos(2*pi*rate1*x1) + amp

start_time <- as.POSIXct("2023-01-01 00:00:00")
end_time <- as.POSIXct("2023-12-31 00:00:00")
timestamp_sequence <- seq.POSIXt(start_time, end_time, by = "144 min")
times <- strftime(timestamp_sequence, "%Y-%m-%dT%H:%M:%SZ")
data <- data.frame(cbind(times,rep("QINE",3650),rep("FAKE1",3650),rep("CMS",3650),y1))
colnames(data) <- c("value_date","variable_name","location","measurement_unit","value")
write.table(data, file="obs.csv", sep=",", row.names = FALSE, quote=FALSE, col.names=T)
```

Next, use this dataset on both sides of the evaluation with default event detection parameters:

```
observed: obs.csv
predicted: obs.csv
event_detection: observed
metrics:
  - sample size
output_formats:
  - csv2
```

This produces the following detected events:

![example1](https://github.com/user-attachments/assets/4915d8b0-76f3-43a0-b994-6d7e720457ae)

Seems somewhat reasonable, I suppose. I mean, it identifies the bumps. The events are a bit lagged and a bit short, but it's reasonable.

But the parameters can be tweaked to lag the events back and broaden them. This is a faff, but it works:

```
observed: obs.csv
predicted: obs.csv
event_detection:
  dataset:  observed
  parameters:
    window_size: 4800
    start_radius: 380
    duration_unit: hours
metrics:
  - sample size
output_formats:
  - csv2
```

Which produces this. There is one fewer event, but this is broadly understandable because the series is curtailed, so if you tweak the detection in this direction, the last event isn't going to appear. Likewise the first.

![example2](https://github.com/user-attachments/assets/1dffdc19-0f9d-4fea-b11b-2e96dc2f7f7b)

Bottom line, the mechanics seem to be working OK and the technique basically works and is somewhat flexible, but it I am unsure about the chance of a user getting the results they might expect with the defaults or, indeed, any initial set of parameter values. Obviously, this challenge only gets bigger when looking across locations because the parameters are declared for all locations, yet their interpretation/meaning obviously depends on the basin-specific characteristic of the time-series being scanned.

That said, I was not optimistic about achieving much more than this. If this new feature generates interest and use, we may want to invest in more sophisticated detection approaches that involve fewer user-defined parameter values, but that is not a trivial amount of work and is definitely not in the first cut.

**Comment by james-d-brown** [comment]
Ah, I was in the process of creating another example, extending the above to add a phase offset to a `predicted` dataset, followed by aggregation of the intersecting observed and predicted events whereby the aggregated event spans the overall period of the intersecting events, like so:

```
observed: obs.csv
predicted: obs.csv
event_detection:
  dataset: 
    - observed
    - predicted
  combination:
      operation: intersection
      aggregation: maximum
metrics:
  - sample size
  - time to peak error
```

This produces the following aggregate events, which looks reasonable:

![example3](https://github.com/user-attachments/assets/da51b33b-abfa-40ac-812c-0ae041469742)

However, it also produces this error, which will need a separate ticket as it's unrelated to event detection. Timing error metrics do not currently support non-forecasts datasets, yet the above will be one of the main applications of event detection, I expect.

```
Caused by: wres.config.yaml.DeclarationException: Encountered 1 error(s) in the declared evaluation, which must be fixed:
    - The declared or inferred data 'type' for the 'predicted' dataset is observations, but the following metrics require single-valued forecasts: [TIME TO PEAK ERROR]. Please remove these metrics or correct the data 'type' to 'single valued forecasts'.
```

**Comment by james-d-brown** [comment]
Will create that ticket tomorrow.

**Comment by james-d-brown** [comment]
Draft wiki is here: https://github.com/NOAA-OWP/wres/wiki/Event-detection

No doubt, it can be improved.

**Comment by james-d-brown** [comment]
There is a failing system test, scenario510, but it is not failing locally. Struggling to find the resources created by the action to debug properly.

**Comment by james-d-brown** [comment]
Rebuilding systests helps for some unknown reason, as nothing has changed w/r to the systests. Possibly an issue with installing the latest graphics. Anyway, can debug now.

**Comment by james-d-brown** [comment]
In writing the wiki (and testing it works!), discovered another issue - the `purpose` of a covariate dataset is not interpolated from the context when available, recalling that a covariate may be used to `filter` or to `detect` events.

There are so many moving parts w/r to how event detection interacts with other features. As noted before, the UAT phase for this will be fairly extensive and I anticipate several issues, despite the decent unit and integration test coverage.

**Comment by james-d-brown** [comment]
Will also need to add some declaration validation with warnings about the assumptions made when the `purpose` of a covariate is undeclared.

**Comment by james-d-brown** [comment]
With the last commit, this branch should be ready to merge into main so that UAT can begin (or, rather, pre-deployment testing).

**Comment by james-d-brown** [comment]
Merge in progress, should then be ready for UAT.

**Comment by james-d-brown** [comment]
Although ready for UAT, I will continue to do some testing of this myself. There are a lot of moving parts and interactions with existing features.

**Comment by james-d-brown** [comment]
Again, the wiki is here:

https://github.com/NOAA-OWP/wres/wiki/Event-detection

**Comment by james-d-brown** [comment]
Pending UAT.

**Comment by james-d-brown** [comment]
Hmm, looks like some system tests failed related to NetCDF reading. I think SQ caught a couple of issues w/r to some code there, which was touched by a class rename, and I accepted the edit without rerunning the system tests, so that is likely the culprit. Will reverse that minor change now...

**Comment by james-d-brown** [comment]
Done.

**Comment by HankHerr-NOAA** [comment]
Reviewing the wiki...

I have a question about the example, 

```yaml
observed: observed.csv
predicted: predicted.csv
covariates:
  - sources: precipitation.csv
    purpose: detect
  - sources: temperature.csv
event_detection: covariates
```
and following paragraph:

> In this case, only the precipitation.csv will be used for event detection. When event_detection is declared and the purpose of a covariate is undeclared, then the context of that covariate is inspected to determine whether it should be used for event detection or filtering. Specifically, if there are no filtering parameters (i.e., no minimum or maximum), then the purpose is assumed to be detect, otherwise filter.

So the `purpose` of the temperature CSV is undeclared. Per your paragraph, this applies:

"Specifically, if there are no filtering parameters (i.e., no minimum or maximum), then the purpose is assumed to be detect,..."

No filtering parameters are declared, so the purpose of the temperature covariate should be `detect`. Yet, it will not be used for event detection according to the first sentence (only precip is). Why is precipitation the only one used for event detection?

I guess I'm not understanding the paragraph. Please let me know where I've gone wrong. Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Need to step away for a bit, but, after returning, I hope to complete my review of the wiki before I leave a doctor's appointment. Back in 30 minutes or so,

Hank

**Comment by james-d-brown** [comment]
Yes, good catch, there is meant to be a filter on temperature, just forgot to add one, will fix.

**Comment by james-d-brown** [comment]
Fixed the declaration and clarified the explanation.

**Comment by HankHerr-NOAA** [comment]
Looks good. Thanks!

Continuing my review,

Hank

**Comment by HankHerr-NOAA** [comment]
I'm almost hesitant to write this in case my math is screwed up, but here it goes... 

The defaults for `window_size` and `half_life` are inconsistent with the rules for computing each when the other is defined. See [this section](https://github.com/NOAA-OWP/wres/wiki/Event-detection#what-are-the-possible-values-of-the-event-detection-parameters).

If neither is defined, then, `window_size` = (avg time step) * 200 and `half_life` = (avg time step) * 20, so `window_size` / `half_life` = 10.

If one of the two are defined, then, when `window_size` is defined, `half_life` = `window_size`/20, and when `half_life` is defined, `window_size` = 20 * `half_life`. So `window_size` / `half_life` = 20.

A ratio of 20 when one is defined vs. 10 when neither is defined.  I'm not sure why there would be that inconsistency, and I doubt anyone would notice unless they really look carefully at that table in that section... which I did.

Regardless, I'm not proposing a change, unless you think there is an error somewhere in either the code or the wiki. Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
I may have overlooked it and need to go soon so I don't have time to double check...

Are the events reported in the evaluation CSV? Are they reported in the logging? It may be good to let the user know how to find those events. The images appear to only provide the earliest valid time (start time of the event?), based on the x-axis label.

Off to the doctor, 

Hank

**Comment by james-d-brown** [comment]
> I'm almost hesitant to write this in case my math is screwed up, but here it goes...
> 
> The defaults for `window_size` and `half_life` are inconsistent with the rules for computing each when the other is defined. See [this section](https://github.com/NOAA-OWP/wres/wiki/Event-detection#what-are-the-possible-values-of-the-event-detection-parameters).
> 
> If neither is defined, then, `window_size` = (avg time step) * 200 and `half_life` = (avg time step) * 20, so `window_size` / `half_life` = 10.
> 
> If one of the two are defined, then, when `window_size` is defined, `half_life` = `window_size`/20, and when `half_life` is defined, `window_size` = 20 * `half_life`. So `window_size` / `half_life` = 20.
> 
> A ratio of 20 when one is defined vs. 10 when neither is defined. I'm not sure why there would be that inconsistency, and I doubt anyone would notice unless they really look carefully at that table in that section... which I did.
> 
> Regardless, I'm not proposing a change, unless you think there is an error somewhere in either the code or the wiki. Thanks,
> 
> Hank

Yeah, we can make the ratio consistent. The defaults are a bit hand-wavey, but we can at least ensure internal consistency.

So: 

W=10H for H declared
H=W/10 or W=10H for W declared
H=20TS, W=200TS or W=10H for neither declared

Will adjust.

**Comment by james-d-brown** [comment]
> I may have overlooked it and need to go soon so I don't have time to double check...
> 
> Are the events reported in the evaluation CSV? Are they reported in the logging? It may be good to let the user know how to find those events. The images appear to only provide the earliest valid time (start time of the event?), based on the x-axis label.
> 
> Off to the doctor,
> 
> Hank

The "events" are just time windows, reported in the same way that time windows are reported more generally, which includes the logging and the various numeric output formats, including pairs. The graphics are just a visual representation of (part of) the information and it's the same for the "event" time windows as any other ("non-event" or, rather, declared) time windows. One thing that's noted in the limitations is that it would make sense to separate out event detection from evaluation. That way, the events are more transparent to a user iterating them.

I could add a sentence about the output formats recording them....?

**Comment by james-d-brown** [comment]
> Off to the doctor,

Also off to a medical appointment now, will be back in an hour or two...

**Comment by HankHerr-NOAA** [comment]
> I could add a sentence about the output formats recording them....?

Yeah, perhaps. No more than a sentence. Thanks!

Hank

**Comment by HankHerr-NOAA** [comment]
The wiki looks good otherwise. I'll start my testing on Tuesday, and, as mentioned in the meeting, I'll use a separate ticket to track that testing since there will be a lot of it.

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Moving this to the just created v6.30 milestone,

Hank
